Support loading Web Extension URLs and load the background page or service worker.

xeenon opened the pull request:
#### 37568d474f72b610d2dd7c356a6709388ff669c1
<pre>
Support loading Web Extension URLs and load the background page or service worker.
<a href="https://bugs.webkit.org/show_bug.cgi?id=246409">https://bugs.webkit.org/show_bug.cgi?id=246409</a>

Reviewed by Brian Weinstein.

This also adds _WKWebExtensionController to WKWebViewConfiguration and propagate through to WebPage.

* Source/WebCore/en.lproj/Localizable.strings: Updated with update-webkit-localizable-strings.
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView urlSchemeHandlerForURLScheme:]): Deleted. Dead code, not used or exposed in any header.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration urlSchemeHandlerForURLScheme:]): Only return API handlers.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(-[_WKWebExtensionContext setBaseURL:]): Added more assertions.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.mm:
(-[_WKWebExtensionController extensionContextForExtension:]): Added baseURL version.
* Source/WebKit/UIProcess/Cocoa/WebURLSchemeHandlerCocoa.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::isAccessibleResourcePath): Added with FIXME.
(WebKit::WebExtension::resourceDataForPath): Remove &quot;/&quot; prefix.
(WebKit::toAPI): Added.
(WebKit::WebExtension::createError): Use toAPI.
(WebKit::WebExtension::removeError): Added.
(WebKit::WebExtension::recordError): Dispatch notifications async to avoid reentrancy.
(WebKit::WebExtension::backgroundContentPath): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(-[_WKWebExtensionContextDelegate initWithWebExtensionContext:]): Added.
(-[_WKWebExtensionContextDelegate webView:decidePolicyForNavigationAction:decisionHandler:]): Added.
(-[_WKWebExtensionContextDelegate webView:didFinishNavigation:]): Added.
(-[_WKWebExtensionContextDelegate webView:didFailNavigation:withError:]): Added.
(-[_WKWebExtensionContextDelegate webViewWebContentProcessDidTerminate:]): Added.
(WebKit::WebExtensionContext::WebExtensionContext): Create m_delegate.
(WebKit::toAPI): Added.
(WebKit::WebExtensionContext::createError): Use toAPI and added BaseURLTaken case.
(WebKit::WebExtensionContext::load): Call loadBackgroundWebViewDuringLoad.
(WebKit::WebExtensionContext::unload): Call unloadBackgroundWebView.
(WebKit::WebExtensionContext::setBaseURL): Canonicalize URL scheme.
(WebKit::WebExtensionContext::postAsyncNotification): Use makeBlockPtr to be safe in C++.
(WebKit::WebExtensionContext::webViewConfiguration): Added.
(WebKit::WebExtensionContext::backgroundContentURL): Added.
(WebKit::WebExtensionContext::loadBackgroundWebViewDuringLoad): Added.
(WebKit::WebExtensionContext::loadBackgroundWebView): Added.
(WebKit::WebExtensionContext::unloadBackgroundWebView): Added.
(WebKit::WebExtensionContext::performTasksAfterBackgroundContentLoads): Added.
(WebKit::WebExtensionContext::decidePolicyForNavigationAction): Added.
(WebKit::WebExtensionContext::didFinishNavigation): Added.
(WebKit::WebExtensionContext::didFailNavigation): Added.
(WebKit::WebExtensionContext::webViewWebContentProcessDidTerminate):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::load): Added scheme handling more more error reporting.
(WebKit::WebExtensionController::unload): Update the baseURL map.
(WebKit::WebExtensionController::addPage): Added. Register the scheme handlers.
(WebKit::WebExtensionController::removePage): Added.
(WebKit::WebExtensionController::extensionContext const):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionURLSchemeHandlerCocoa.mm: Added.
(WebKit::WebExtensionURLSchemeHandler::WebExtensionURLSchemeHandler):
(WebKit::WebExtensionURLSchemeHandler::platformStartTask):
(WebKit::WebExtensionURLSchemeHandler::platformStopTask):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
(WebKit::WebExtensionController::wrapper const):
* Source/WebKit/UIProcess/Extensions/WebExtensionURLSchemeHandler.h: Added.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::WebPageProxy): Call addPage on the WebExtensionController.
(WebKit::WebPageProxy::close): Call removePage on the WebExtensionController.
* Source/WebKit/UIProcess/WebURLSchemeHandler.h:
(WebKit::WebURLSchemeHandler::isAPIHandler): Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Added new files.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionController.mm:
(TestWebKitAPI::TEST(WKWebExtensionController, BackgroundPageLoading)): Added.
* Source/WebKit/Shared/WebPageCreationParameters.cpp:
(WebKit::WebPageCreationParameters::encode const): Encode WebExtensionControllerParameters.
(WebKit::WebPageCreationParameters::decode): Decode WebExtensionControllerParameters.
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/UIProcess/API/APIPageConfiguration.cpp:
(API::PageConfiguration::copy const): Copy _webExtensionController.
(API::PageConfiguration::webExtensionController): Added.
(API::PageConfiguration::setWebExtensionController): Added.
* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setupPageConfiguration:]): Copy _webExtensionController.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration encodeWithCoder:]): Encode _webExtensionController.
(-[WKWebViewConfiguration initWithCoder:]): Decode _webExtensionController.
(-[WKWebViewConfiguration copyWithZone:]): Copy _webExtensionController.
(-[WKWebViewConfiguration _webExtensionController]): Added.
(-[WKWebViewConfiguration _setWebExtensionController:]): Added.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationPrivate.h:
Added _webExtensionController property.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::WebPageProxy): Set m_webExtensionController.
(WebKit::WebPageProxy::creationParameters): Set webExtensionControllerParameters.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: <a href="https://commits.webkit.org/255512@main">https://commits.webkit.org/255512@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/16ec4d3454bcb95f3a3801cbbebde19a08f821e0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92779 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1992 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/23363 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/102510 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/162826 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/96781 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1993 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/30338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/85183 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/98656 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98442 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/1353 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/79267 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/85183 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/83267 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/23363 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/85183 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/36740 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/23363 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/34539 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/23363 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/63/builds/38409 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/79267 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1750 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/65/builds/40323 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/23363 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
```diff
@@ -297,6 +300,11 @@ - (void)encodeWithCoder:(NSCoder *)coder
 
+#if ENABLE(WK_WEB_EXTENSIONS)
+    if (auto *controller = _webExtensionController.peek())
+        [coder encodeObject:controller forKey:@"webExtensionController"];
```
It looks like we don't put these keys in constants anywhere else.

b-weinstein replied:
(I was going to say maybe we should, but it wouldn't be consistent)

xeenon replied:
Yeah, less of that in WebKit.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -240,6 +241,12 @@
 
+bool WebExtension::isAccessibleResourcePath(NSString *, NSURL *pageURL)
+{
+    // FIXME: Implement web accessible resources.
```
Nit: Is there a bug number? Or do we not include those here.

xeenon replied:
I'll file some bugs for these `FIXME`s.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -312,24 +323,65 @@
 NSError *WebExtension::createError(Error error, NSString *customLocalizedDescription, NSError *underlyingError)
 {
-    _WKWebExtensionError errorCode;
+    auto errorCode = toAPI(error);
```
I like this refactoring!

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -44,10 +46,34 @@
+        return false;
+    }
+
+    if (!extensionContext.load(*this, outError)) {
```
Do we want to do anything with `outError`? Or would any error be logged/vended elsewhere?

xeenon replied:
It is just returned to the API call site, and is the primary way Swift handles errors. These turn into a `throws` `func` in Swift.

xeenon committed:
Support loading Web Extension URLs and load the background page or service worker.
https://bugs.webkit.org/show_bug.cgi?id=246409

Reviewed by Brian Weinstein.

This also adds _WKWebExtensionController to WKWebViewConfiguration and propagate through to WebPage.

* Source/WebCore/en.lproj/Localizable.strings: Updated with update-webkit-localizable-strings.
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView urlSchemeHandlerForURLScheme:]): Deleted. Dead code, not used or exposed in any header.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration urlSchemeHandlerForURLScheme:]): Only return API handlers.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(-[_WKWebExtensionContext setBaseURL:]): Added more assertions.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.mm:
(-[_WKWebExtensionController extensionContextForExtension:]): Added baseURL version.
* Source/WebKit/UIProcess/Cocoa/WebURLSchemeHandlerCocoa.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::isAccessibleResourcePath): Added with FIXME.
(WebKit::WebExtension::resourceDataForPath): Remove "/" prefix.
(WebKit::toAPI): Added.
(WebKit::WebExtension::createError): Use toAPI.
(WebKit::WebExtension::removeError): Added.
(WebKit::WebExtension::recordError): Dispatch notifications async to avoid reentrancy.
(WebKit::WebExtension::backgroundContentPath): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(-[_WKWebExtensionContextDelegate initWithWebExtensionContext:]): Added.
(-[_WKWebExtensionContextDelegate webView:decidePolicyForNavigationAction:decisionHandler:]): Added.
(-[_WKWebExtensionContextDelegate webView:didFinishNavigation:]): Added.
(-[_WKWebExtensionContextDelegate webView:didFailNavigation:withError:]): Added.
(-[_WKWebExtensionContextDelegate webViewWebContentProcessDidTerminate:]): Added.
(WebKit::WebExtensionContext::WebExtensionContext): Create m_delegate.
(WebKit::toAPI): Added.
(WebKit::WebExtensionContext::createError): Use toAPI and added BaseURLTaken case.
(WebKit::WebExtensionContext::load): Call loadBackgroundWebViewDuringLoad.
(WebKit::WebExtensionContext::unload): Call unloadBackgroundWebView.
(WebKit::WebExtensionContext::setBaseURL): Canonicalize URL scheme.
(WebKit::WebExtensionContext::postAsyncNotification): Use makeBlockPtr to be safe in C++.
(WebKit::WebExtensionContext::webViewConfiguration): Added.
(WebKit::WebExtensionContext::backgroundContentURL): Added.
(WebKit::WebExtensionContext::loadBackgroundWebViewDuringLoad): Added.
(WebKit::WebExtensionContext::loadBackgroundWebView): Added.
(WebKit::WebExtensionContext::unloadBackgroundWebView): Added.
(WebKit::WebExtensionContext::performTasksAfterBackgroundContentLoads): Added.
(WebKit::WebExtensionContext::decidePolicyForNavigationAction): Added.
(WebKit::WebExtensionContext::didFinishNavigation): Added.
(WebKit::WebExtensionContext::didFailNavigation): Added.
(WebKit::WebExtensionContext::webViewWebContentProcessDidTerminate):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::load): Added scheme handling more more error reporting.
(WebKit::WebExtensionController::unload): Update the baseURL map.
(WebKit::WebExtensionController::addPage): Added. Register the scheme handlers.
(WebKit::WebExtensionController::removePage): Added.
(WebKit::WebExtensionController::extensionContext const):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionURLSchemeHandlerCocoa.mm: Added.
(WebKit::WebExtensionURLSchemeHandler::WebExtensionURLSchemeHandler):
(WebKit::WebExtensionURLSchemeHandler::platformStartTask):
(WebKit::WebExtensionURLSchemeHandler::platformStopTask):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
(WebKit::WebExtensionController::wrapper const):
* Source/WebKit/UIProcess/Extensions/WebExtensionURLSchemeHandler.h: Added.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::WebPageProxy): Call addPage on the WebExtensionController.
(WebKit::WebPageProxy::close): Call removePage on the WebExtensionController.
* Source/WebKit/UIProcess/WebURLSchemeHandler.h:
(WebKit::WebURLSchemeHandler::isAPIHandler): Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Added new files.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionController.mm:
(TestWebKitAPI::TEST(WKWebExtensionController, BackgroundPageLoading)): Added.
* Source/WebKit/Shared/WebPageCreationParameters.cpp:
(WebKit::WebPageCreationParameters::encode const): Encode WebExtensionControllerParameters.
(WebKit::WebPageCreationParameters::decode): Decode WebExtensionControllerParameters.
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/UIProcess/API/APIPageConfiguration.cpp:
(API::PageConfiguration::copy const): Copy _webExtensionController.
(API::PageConfiguration::webExtensionController): Added.
(API::PageConfiguration::setWebExtensionController): Added.
* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setupPageConfiguration:]): Copy _webExtensionController.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration encodeWithCoder:]): Encode _webExtensionController.
(-[WKWebViewConfiguration initWithCoder:]): Decode _webExtensionController.
(-[WKWebViewConfiguration copyWithZone:]): Copy _webExtensionController.
(-[WKWebViewConfiguration _webExtensionController]): Added.
(-[WKWebViewConfiguration _setWebExtensionController:]): Added.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationPrivate.h:
Added _webExtensionController property.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::WebPageProxy): Set m_webExtensionController.
(WebKit::WebPageProxy::creationParameters): Set webExtensionControllerParameters.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: https://commits.webkit.org/255512@main
